cli/ops/websocket.rs

```diff
@@ -229,33 +233,23 @@ pub async fn op_ws_close(
 ) -> Result<Value, AnyError> {
   let args: CloseArgs = serde_json::from_value(args)?;
   let rid = args.rid;
-  let mut maybe_msg = Some(Message::Close(args.code.map(|c| CloseFrame {
+  let msg = Message::Close(args.code.map(|c| CloseFrame {
     code: CloseCode::from(c),
     reason: match args.reason {
       Some(reason) => Cow::from(reason),
       None => Default::default(),
     },
-  })));
-
-  poll_fn(move |cx| {
-    let mut state = state.borrow_mut();
-    let stream = state
-      .resource_table
-      .get_mut::<WsStream>(rid)
-      .ok_or_else(bad_resource_id)?;
-
-    // TODO(ry) Handle errors below instead of unwrap.
-    // Need to map `TungsteniteError` to `AnyError`.
-    ready!(stream.poll_ready_unpin(cx)).unwrap();
-    if let Some(msg) = maybe_msg.take() {
-      stream.start_send_unpin(msg).unwrap();
-    }
-    ready!(stream.poll_flush_unpin(cx)).unwrap();
-    ready!(stream.poll_close_unpin(cx)).unwrap();
-
-    Poll::Ready(Ok(json!({})))
-  })
-  .await
+  }));
+
+  let resource = state
+    .borrow_mut()
+    .resource_table_2
+    .get::<WsStreamResource>(rid)
+    .ok_or_else(bad_resource_id)?;
+  let mut ws_stream = resource.borrow_mut().await;
+  ws_stream.send(msg).await?;
+  ws_stream.close(None).await?;
+  Ok(json!({}))
```

Comment:
either remove `ws_stream.close()`, or make `msg` not `Message::Close`, but just the `Option<CloseFrame>` and call `ws_stream.close()` with `msg` and delete `ws_stream.send()`. `close()` is just a wrapper around send, so what you are doing is closing twice, once with a optional closeframe and once with no close frame (which just defaults internally to a closeframe)

Reply:
Can I just remove `ws_stream.close()`?

Reply:
yes